### DISCUSSION TITLE
Scheduled deployment

### DISCUSSION THREAD

**Discussion by azmeuk** [woodpecker-ci/woodpecker#2281] [Request a Feature] [answered]
### Clear and concise description of the problem

CI can be long to run, and it must pass before the deploy button is displayed. So users who want to deploy something have to come back regularly to woodpecker to check if the CI have finished.

### Suggested solution

I suggest that, when the CI is still running, a "Deploy on success" button should be displayed. This way users could just hit that button and go do what they want, knowing that deployment will happen if the CI is green.

### Additional context

This is inspired from the Gitlab "Merge on success" button that is displayed on merge requests when the CI is running.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by lafriks** [comment]
This can be already be done by adding step using woodpeckerci/plugin-trigger to trigger deployment task in same/other repositories

See https://woodpecker-ci.org/plugins/Trigger

**Comment by azmeuk** [comment]
I understand that this plugin could be used for automated deployment, but I am not sure it would cover this usecase.

In the scenario I describe, I expect users to manually hit the `deploy` button if they want to deploy. This is just that the button would be displayed before the CI is ended.

**Comment by 6543** [comment]
it's an edge-case feature request ... normally you either are confident in the pipeline checks and let it deploy automatically ... or you deploy manually

deploy manually as it would for a auto-deploy is ... strange

I personally thing that this feature just add complexity and don't encourage good practices ... so my vote goes against adding this ...

**Comment by 6543** [comment]
cc @woodpecker-ci/maintainers :point_up:

**Comment by azmeuk** [comment]
Thank you for your answer. I will try to give a little more context.
I have projects with several deployment targets (production, testing, production-2, etc.), and on which I need to deploy different branches. The `main` branch is generally deployed to `testing`, but not every commit, and sometimes I deploy feature branches on `testing`. I generally deploy a `production` branch on the targets `production` and `production2`, but sometimes I deploy other stable branches on `production2`. And this varies depending on the projects.

I understand that you suggest an automatic deploy strategy. This would be for instance, having one branch for each deployment target, and automatically deploy those branches on each new commit.

However I think that would not be flexible enough for me. As I generally don't want to deploy each commit, this would force me to add a lot more branches to my workflow, just for the sake of deployment. I fear this would be a hell of merges and `git reset --hard` .

**Comment by anbraten** [comment]
Its an interesting feature, but the use case sounds pretty extraordinary to me as well. 😅

You either want it to be manual or automatic IMO. Why isn't using the manual workflow an option for you? You run basic checks on push automatically and if that is successful you can manually trigger a job for publishing/ releasing and define your target environment for it.

As you want to select an environment anyways you always need the manual step.

**Comment by azmeuk** [comment]
> Why isn't using the manual workflow an option for you?

I am not sure how the manual workflow would work with my usecase. :thinking: 

1. push some code
2. unit test pipeline is triggered
3. trigger manual workflow, choose a deployment environment
4. unit test pipeline is triggered again (maybe cancel the first run?)
5. deployment pipeline is triggered if 4. suceeds

Is this what you are thinking about?

> As you want to select an environment anyways you always need the manual step.

Indeed. The idea is not to remove that unavoidable manual step, it is just to allow users to do it early.

At the moment, each time I want to deploy, I have to wait for the unit tests to finish and pass, and come back to woodpecker to trigger the deployment. I would love to be able to schedule the deployment when the CI is running, go to do other stuff and not having to get back to woodpecker with the hope CI is finished, in order to trigger the deployment.

This is similar to the [gitlab 'merge whene pipeline succeeds'](https://docs.gitlab.com/ee/user/project/merge_requests/merge_when_pipeline_succeeds.html) feature I often use on gitlab:

![mwps_v15_4](https://user-images.githubusercontent.com/60163/225910822-e537d61e-8d07-46c8-a593-d1883d5f6132.png)

**Comment by qwerty287** [comment, marked as answer]
As we explained, this should be possible with the trigger plugin (which auto-triggers a manual pipeline). Otherwise that's a kinda rare usecase, and I don't think we can support every small change.

Thus, I'm closing this.
